### PR TITLE
Add hid commands for setting led matrix mode

### PR DESCRIPTION
**Pull request by ids1024**
This works, but we probably want to restrict which modes can be set, and I'm not sure if there's any possibility for the numbering of the modes to change.

### CODE REVIEW

File: keyboards/system76/system76_ec.c [hunk 1]
```diff
@@ -172,6 +174,16 @@ void raw_hid_receive(uint8_t *data, uint8_t length) {
                 }
             }
             break;
+        case CMD_LED_GET_MODE:
+            data[2] = rgb_matrix_get_mode();
```
Comment (jackpot51) [inline keyboards/system76/system76_ec.c:178]: I blocked get when unlocked too, since it did not make much sense to read LED information at that time

**Comment by ids1024** [comment]
> I'm not sure if there's any possibility for the numbering of the modes to change.

It seems my concern on this was justified:

```rust
enum rgb_matrix_effects {
    RGB_MATRIX_NONE = 0,
    RGB_MATRIX_SOLID_COLOR = 1,     // Static single hue, no speed support
    RGB_MATRIX_ALPHAS_MODS,         // Static dual hue, speed is hue for secondary hue
    RGB_MATRIX_GRADIENT_UP_DOWN,    // Static gradient top to bottom, speed controls how much gradient changes
    RGB_MATRIX_GRADIENT_LEFT_RIGHT,    // Static gradient left to right, speed controls how much gradient changes
    RGB_MATRIX_BREATHING,           // Single hue brightness cycling animation
    RGB_MATRIX_BAND_SAT,        // Single hue band fading saturation scrolling left to right
    RGB_MATRIX_BAND_VAL,        // Single hue band fading brightness scrolling left to right
    RGB_MATRIX_BAND_PINWHEEL_SAT,   // Single hue 3 blade spinning pinwheel fades saturation
    RGB_MATRIX_BAND_PINWHEEL_VAL,   // Single hue 3 blade spinning pinwheel fades brightness
    RGB_MATRIX_BAND_SPIRAL_SAT,     // Single hue spinning spiral fades saturation
    RGB_MATRIX_BAND_SPIRAL_VAL,     // Single hue spinning spiral fades brightness
    RGB_MATRIX_CYCLE_ALL,           // Full keyboard solid hue cycling through full gradient
    RGB_MATRIX_CYCLE_LEFT_RIGHT,    // Full gradient scrolling left to right
    RGB_MATRIX_CYCLE_UP_DOWN,       // Full gradient scrolling top to bottom
    RGB_MATRIX_CYCLE_OUT_IN,        // Full gradient scrolling out to in
    RGB_MATRIX_CYCLE_OUT_IN_DUAL,   // Full dual gradients scrolling out to in
    RGB_MATRIX_RAINBOW_MOVING_CHEVRON,  // Full gradent Chevron shapped scrolling left to right
    RGB_MATRIX_CYCLE_PINWHEEL,      // Full gradient spinning pinwheel around center of keyboard
    RGB_MATRIX_CYCLE_SPIRAL,        // Full gradient spinning spiral around center of keyboard
    RGB_MATRIX_DUAL_BEACON,         // Full gradient spinning around center of keyboard
    RGB_MATRIX_RAINBOW_BEACON,      // Full tighter gradient spinning around center of keyboard
    RGB_MATRIX_RAINBOW_PINWHEELS,   // Full dual gradients spinning two halfs of keyboard
    RGB_MATRIX_RAINDROPS,           // Randomly changes a single key's hue
    RGB_MATRIX_JELLYBEAN_RAINDROPS, // Randomly changes a single key's hue and saturation
#if define(RGB_MATRIX_FRAMEBUFFER_EFFECTS)
    RGB_MATRIX_TYPING_HEATMAP,      // How hot is your WPM!
    RGB_MATRIX_DIGITAL_RAIN,        // That famous computer simulation
#endif
#if defined(RGB_MATRIX_KEYPRESSES) || defined(RGB_MATRIX_KEYRELEASES)
    RGB_MATRIX_SOLID_REACTIVE_SIMPLE,   // Pulses keys hit to hue & value then fades value out
    RGB_MATRIX_SOLID_REACTIVE,      // Static single hue, pulses keys hit to shifted hue then fades to current hue
    RGB_MATRIX_SOLID_REACTIVE_WIDE       // Hue & value pulse near a single key hit then fades value out
    RGB_MATRIX_SOLID_REACTIVE_MULTIWIDE  // Hue & value pulse near multiple key hits then fades value out
    RGB_MATRIX_SOLID_REACTIVE_CROSS      // Hue & value pulse the same column and row of a single key hit then fades value out
    RGB_MATRIX_SOLID_REACTIVE_MULTICROSS // Hue & value pulse the same column and row of multiple key hits then fades value out
    RGB_MATRIX_SOLID_REACTIVE_NEXUS      // Hue & value pulse away on the same column and row of a single key hit then fades value out
    RGB_MATRIX_SOLID_REACTIVE_MULTINEXUS // Hue & value pulse away on the same column and row of multiple key hits then fades value out
    RGB_MATRIX_SPLASH,              // Full gradient & value pulse away from a single key hit then fades value out
    RGB_MATRIX_MULTISPLASH,         // Full gradient & value pulse away from multiple key hits then fades value out
    RGB_MATRIX_SOLID_SPLASH,        // Hue & value pulse away from a single key hit then fades value out
    RGB_MATRIX_SOLID_MULTISPLASH,   // Hue & value pulse away from multiple key hits then fades value out
#endif
    RGB_MATRIX_EFFECT_MAX
};
```

(The actual code seems to use include and macro magic to expand to something like this).

So the numbering depends on what defines are set. I guess if we change those in our firmware, we would have to make changes to this part of QMK to make sure the numbering stays the same.

Don't know if there's a better way to handle it.

File: keyboards/system76/system76_ec.c [hunk 2]
```diff
@@ -161,13 +200,30 @@ void raw_hid_receive(uint8_t *data, uint8_t length) {
...
+            }
+            break;
+        case CMD_LED_GET_MODE:
+            if (!bootloader_unlocked) {
+                enum rgb_matrix_effects mode = rgb_matrix_get_mode();
+                for (uint8_t i = 0; i < MODE_LAST; i++) {
+                    if (mode_map[i] == mode) {
+                        data[2] = mode;
+                        data[3] = rgb_matrix_config.speed;
+                        data[1] = 0;
```
Comment (jackpot51) [inline keyboards/system76/system76_ec.c:217]: `break;` after this line

File: keyboards/system76/system76_ec.c [hunk 3]
```diff
@@ -161,13 +200,30 @@ void raw_hid_receive(uint8_t *data, uint8_t length) {
                 };
                 if (index < DRIVER_LED_TOTAL) {
                     raw_rgb_data[index] = rgb;
-                    rgb_matrix_mode_noeeprom(RGB_MATRIX_CUSTOM_raw_rgb);
                     data[1] = 0;
                 } else if (index == CMD_LED_INDEX_ALL) {
-                    for (int i = 0; i < DRIVER_LED_TOTAL; i++) {
-                        raw_rgb_data[i] = rgb;
+                    rgb_matrix_sethsv_noeeprom(rgb.r, rgb.g, rgb_matrix_config.hsv.v);
```
Comment (jackpot51) [inline keyboards/system76/system76_ec.c:205]: Also adjust the LED_GET_COLOR code

**Review by jackpot51 — APPROVED** [review]

**Pull request merged**